libs/functional/tests/unit/tag_invoke.cpp

```diff
@@ -13,23 +13,44 @@
 namespace mylib {
     HPX_INLINE_CONSTEXPR_VARIABLE struct foo_fn
     {
+        // Note the use of foo_fn() instead of *this. Not all compilers (i.e.
+        // gcc <= 8) are compliant enough to allow *this, and would otherwise
+        // report "invalid use of 'this' at top level". However, clang doesn't
+        // support using foo_fn() and reports "invalid use of incomplete type
+        // 'mylib::foo_fn'" so we still use *this for clang and all other
+        // compilers as *this is allowed by the standard.
         template <typename T>
-        constexpr auto operator()(T&& x) const noexcept(
-            noexcept(hpx::functional::tag_invoke(*this, std::forward<T>(x))))
-            -> decltype(hpx::functional::tag_invoke(*this, std::forward<T>(x)))
+        constexpr auto operator()(T&& x) const 
+        // HPX_GCC_VERSION is also defined when clang uses libstdc++. In that
+        // case we still want to use *this.
+#if defined(HPX_GCC_VERSION) && HPX_GCC_VERSION < 90000 && !defined(HPX_CLANG_VERSION)
+            noexcept(noexcept(hpx::functional::tag_invoke(foo_fn(), std::forward<T>(x))))
```

Comment:
@sithhell I'd appreciate your eyes on this one. As far as I can tell using `foo_fn()` is equivalent to `*this` even considering the `const` on `operator()`. But I'm not 100% sure I'm not missing some subtle change that this might cause.

Reply:
`noexcept()` is an unevaluated context, wouldn't `std::declval<foo_fn>()` (or similar) work in any case?

Reply:
Ah, much better, thanks! That seems to work nicely for both clang and gcc.